### PR TITLE
Avoid having to go back to the UI thread in the navbar code

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
@ToddGrun ptal

File: src/EditorFeatures/Core/NavigationBar/NavigationBarController.cs [hunk 1]
```diff
@@ -161,26 +183,6 @@ void IDisposable.Dispose()
         _cancellationTokenSource.Cancel();
     }
 
-    private void Pause()
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/NavigationBar/NavigationBarController.cs:164]: inlined these.

File: src/EditorFeatures/Core/NavigationBar/NavigationBarController.cs [hunk 2]
```diff
@@ -200,31 +202,46 @@ private void StartModelUpdateAndSelectedItemUpdateTasks()
...
         _threadingContext.ThrowIfNotOnUIThread();
-        StartSelectedItemUpdateTask();
+
+        var caretPoint = GetCaretPoint();
+        if (caretPoint == null)
+            return;
+
+        // Cancel any in flight work.  We're on the UI thread, so we know this is the latest position of the user, and that
+        // this should supersede any other selection work items.
+        _selectItemQueue.AddWork(caretPoint.Value, cancelExistingWork: true);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/NavigationBar/NavigationBarController.cs:212]: we now grab the caret and pass it along.

File: src/EditorFeatures/Core/NavigationBar/NavigationBarController_ModelComputation.cs [hunk 3]
```diff
@@ -94,55 +98,27 @@ internal partial class NavigationBarController
...
-        // Cancel any in flight work.  This way we don't update until a short lull after the last user event we received.
-        _selectItemQueue.AddWork(cancelExistingWork: true);
-    }
-
-    private async ValueTask SelectItemAsync(CancellationToken cancellationToken)
+    private async ValueTask SelectItemAsync(ImmutableSegmentedList<int> positions, CancellationToken cancellationToken)
     {
-        // Switch to the UI so we can determine where the user is and determine the state the last time we updated
-        // the UI.
-        await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken).NoThrowAwaitable();
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/NavigationBar/NavigationBarController_ModelComputation.cs:110]: no more need to switch here.

File: src/EditorFeatures/Core/NavigationBar/NavigationBarController_ModelComputation.cs [hunk 4]
```diff
@@ -94,55 +98,27 @@ internal partial class NavigationBarController
...
-
-        var currentSelectedItem = ComputeSelectedTypeAndMember(model, position, cancellationToken);
+        var model = await modelTask.ConfigureAwait(false);
+        var currentSelectedItem = ComputeSelectedTypeAndMember(model, lastCaretPosition, cancellationToken);
 
-        GetProjectItems(out var projectItems, out var selectedProjectItem);
+        var (projectItems, selectedProjectItem) = GetProjectItems();
         if (Equals(model, lastPresentedInfo.model) &&
             Equals(currentSelectedItem, lastPresentedInfo.selectedInfo) &&
             Equals(selectedProjectItem, lastPresentedInfo.selectedProjectItem) &&
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/NavigationBar/NavigationBarController_ModelComputation.cs:124]: in the common case (caret moving in the same member) we will bail out immediately, and not hit hte ui thread.  we only need to go back if we're actually updating the UI thread.

File: src/EditorFeatures/Core/NavigationBar/NavigationBarController_ModelComputation.cs [hunk 5]
```diff
@@ -94,55 +98,27 @@ internal partial class NavigationBarController
...
+        var lastCaretPosition = positions.Last();
 
-        await SelectItemWorkerAsync(cancellationToken).ConfigureAwait(true);
+        // Can grab this directly here as only this queue ever reads or writes to it.
+        var lastPresentedInfo = _lastPresentedInfo;
 
-        // Once we've computed and selected the latest navbar items, pause ourselves if we're no longer visible.
-        // That way we don't consume any machine resources that the user won't even notice.
-        if (_visibilityTracker?.IsVisible(_subjectBuffer) is false)
-            Pause();
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/NavigationBar/NavigationBarController_ModelComputation.cs:122]: this was intentionally removed.  we don't need to check this here for visibility.  We just listen for visbility changes directly and pause/unpause us accordingly.

File: src/EditorFeatures/Core/NavigationBar/NavigationBarController.cs [hunk 6]
```diff
@@ -200,31 +202,46 @@ private void StartModelUpdateAndSelectedItemUpdateTasks()
...
+        var projectItems = documents
+            .Select(d => new NavigationBarProjectItem(
                 d.Project.Name,
                 d.Project.GetGlyph(),
                 workspace: d.Project.Solution.Workspace,
                 documentId: d.Id,
-                language: d.Project.Language)).OrderBy(projectItem => projectItem.Text)];
+                language: d.Project.Language))
+            .OrderBy(projectItem => projectItem.Text)
+            .ToImmutableArray();
```
Comment (ToddGrun) [inline src/EditorFeatures/Core/NavigationBar/NavigationBarController.cs:237]: >ToImmutableArray [](http://example.com/codeflow?start=13&length=16)

nice catch here. One of Linq's best abilities is hiding extra sorts.

File: src/EditorFeatures/Core/NavigationBar/NavigationBarController_ModelComputation.cs [hunk 7]
```diff
@@ -94,55 +98,27 @@ internal partial class NavigationBarController
...
-    }
+        // Make a task that waits indefinitely, or until the cancellation token is signaled.
+        var cancellationTriggeredTask = Task.Delay(-1, cancellationToken);
```
Comment (ToddGrun) [inline src/EditorFeatures/Core/NavigationBar/NavigationBarController_ModelComputation.cs:109]: >Task.Delay(-1, cancellationToken); [](http://example.com/codeflow?start=40&length=34)

Weird, this seems like a common enough desire that I would have thought there would be something like a CancellationToken.ToTask() method or equivalent somewhere to hide the backing Task.Delay(-1, ...).

**Comment by ToddGrun** [comment]
I'm very happy with this change. I feel like mainthread switches are an area that we've neglected in our perf efforts.

**Review by ToddGrun — APPROVED** [review]
:shipit:

**Pull request merged**